src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionsState.cs

```diff
@@ -74,25 +76,87 @@ public SubscriptionConnectionsState(long subscriptionId, SubscriptionStorage sto
+                return;
+            }
+
+            DocumentDatabase.ForTestingPurposes?.Subscription_ActionToCallDuringWaitForSubscribe?.Invoke(_connections);
+
+            connection.AddToStatusDescription("Starting to subscribe.");
+            var sp = Stopwatch.StartNew();
+            while (await _subscriptionConnectingLock.WaitAsync(SubscriptionConnection.WaitForChangedDocumentsTimeoutInMs) == false)
+            {
+                connection.CancellationTokenSource.Token.ThrowIfCancellationRequested();
+                await connection.SendHeartBeatAsync($"A connection from IP '{connection.ClientUri}' is waiting for other concurrent connections to subscribe.");
+                sp.Restart();
+            }
+
+            try
             {
-                using (var old = _disposableNotificationsRegistration)
+                if (_subscriptionState == null)
                 {
-                    _disposableNotificationsRegistration = RegisterForNotificationOnNewDocuments(connection);
+                    // this connection is the first one, we initialize everything
+                    RefreshFeatures(connection);
+                    return;
                 }
 
-                InitializeLastChangeVectorSent(connection.SubscriptionState.ChangeVectorForNextBatchStartingPoint);
-                PreviouslyRecordedChangeVector = LastChangeVectorSent;
+                if (connection.SubscriptionState.RaftIndex < _subscriptionState.RaftIndex)
+                {
+                    // this connection was modified while waiting to subscribe, lets try to drop it
+                    DropSingleConnection(connection, new SubscriptionClosedException($"The subscription '{_subscriptionName}' was modified, connection have to be restarted.", canReconnect: true));
+                    return;
+                }
+
+                if (connection.SubscriptionState.RaftIndex == _subscriptionState.RaftIndex)
+                {
+                    // no changes in the subscription
+                    return;
+                }
+
+                if (connection.SubscriptionState.RaftIndex > _subscriptionState.RaftIndex)
+                {
+                    // we have new connection after subscription have changed
+                    // we have to wait until old connections (with smaller raft index) will get disconnected
+                    // then we continue and will re-initialize 
+                    while (_connections.Any(c => c.SubscriptionState.RaftIndex == _subscriptionState.RaftIndex))
+                    {
+                        connection.CancellationTokenSource.Token.ThrowIfCancellationRequested();
+                        await Task.Delay(300);
+                        await connection.SendHeartBeatIfNeededAsync(sp, $"A connection from IP '{connection.ClientUri}' is waiting for old subscription connections to disconnect.");
+                    }
+
+                    RefreshFeatures(connection);
+                }
```

Comment:
I wonder if instead of all this we should use
`_subscriptions.AddOrUpdate` in the `OpenSubscription` method..

Reply:
All this code happens after `SubscribeAsync` when you already called `OpenSubscriptionAsync` (so `_subscriptions` already was populated with the `SubscriptionConnectionsState`)

Reply:
I mean to move all this logic to the `Update` delegate of the `_subscriptions`